# The Modern Promethease

Please see the [disclaimers](https://github.com/ThomasKAtkins/TheModernPromethease/blob/main/disclaimer.md).

The Modern Promethease is a free and open source tool that replicates the functionality of the tool [Promethease](https://promethease.com/). This tool creates a sumamry of a user's data from the popular genotyping service 23andMe. Example output for the user [Lilly Mendel](https://www.snpedia.com/index.php/User:Lilly_Mendel) is provided in the `mendelgenome/` folder.

## Installation Instructions

```
cd <path where you want to install The Modern Promethease>
git clone https://github.com/ThomasKAtkins/TheModernPromethease.git
cd TheModernPromethease
```

## Running Instructions (using pre-built datasets)

Assumes Python3, Pandas and Numpy are installed. 

Download your raw data from 23andMe by following [these instructions](https://customercare.23andme.com/hc/en-us/articles/212196868-Accessing-Your-Raw-Genetic-Data). Then, edit the provided file by deleting every line that starts with `#` except for the line 

```
# rsid	chromosome	position	genotype
```

which should be changed to:

```
rsid	chromosome	position	genotype
```

Now, to generate a variant report (similar to the one generated by Promethease), run the command

```
python3 generate_variant_report.py <path to 23andMe file>
```

to generate two files, `variant_report.html`, and `snpedia_data.csv`. Opening `variant_report.html` will display a graphical report of the 23andMe data, while `snpedia_data.csv` contains the data in tabular format.

To generate a report containing risk of different conditions, run

```
python3 generate_variant_report.py <path to 23andMe file> <population>
```

where population is one of Global, European, African, AfricanOthers, AfricanAmerican, Asian, EastAsian, OtherAsian, LatinAmerican1, LatinAmerican2, SouthAsian, Other. See the [ALFA Allele Frequencies page](https://www.ncbi.nlm.nih.gov/snp/docs/gsr/data_inclusion/#population) for more information on these categories. The output can be found in `trait_report.html`.


## Re-Generating the Datasets (advanced users only)

One advantage of this tool over Promethease is the ability to update the datasets used to create the tool. We provide somewhat up-to-date versions of these datasets in the `data/` folder, but also provide the code to generate these files with the script `data/getSnpediaPages.R` (requires the `devtools`, `dplyr`, and `readr` libraries). The datasets can be generated using

```
cd data
rm snp_df.csv && rm geno_df.csv
Rscript getSnpediaPages.R <path to 23andMe file>
```

The 23andMe file is necessary so the R script knows which SNPedia pages to scrape. This will produce two new data files, `snp_df.csv` (contains data on the SNP pages) and `geno_df.csv` (contains data on the individual genotype pages).

To generate the data for the trait report, run the `gettraits.py` and `getlinkage.py`. Note: for `getlinkage.py`, an LDLink API token is required.